### PR TITLE
nixos/networking: fix setting MAC Address and MTU in networkd, fix tests

### PR DESCRIPTION
**Pull request by flokli**
###### Motivation for this change
This cleans up code setting MTU and MAC Address. We only use `network-link-${interfaceName}` in scripted networking, and use the native networkd methods to set it in case of networkd.

This also improves the reliability of the networking tests, and fixes `nixosTests.networking.networkd.virtual`.

@rnhmjoj We can just backport a501abd5499d8f82f0991a7b78bcbc4169b0537f into 20.03 to fix the test there as well.

Ran all tests in `nixosTests.networking.networkd` and `nixosTests.networking.scripted` in master, and 20.03 with a501abd5499d8f82f0991a7b78bcbc4169b0537f cherry-picked, they succeed now :tada: 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by teto — APPROVED** [review]

File: nixos/tests/networking.nix [hunk 1]
```diff
@@ -486,7 +493,11 @@ let
             """.format(
                 list, targetList
             )
-
+        with subtest("Test MTU and MAC Address are configured"):
+            assert "mtu 1342" in machine.succeed("ip link show dev tap0")
+            assert "mtu 1343" in machine.succeed("ip link show dev tun0")
+            assert "02:de:ad:be:ef:01" in machine.succeed("ip link show dev tap0")
+        '' + optionalString (!networkd) ''
```
Comment (rnhmjoj) [inline nixos/tests/networking.nix:500]: Can you add a comment here to explain why it's disabled?

File: nixos/modules/tasks/network-interfaces-systemd.nix [hunk 2]
```diff
@@ -94,7 +94,12 @@ in
           address = forEach (interfaceIps i)
             (ip: "${ip.address}/${toString ip.prefixLength}");
           networkConfig.IPv6PrivacyExtensions = "kernel";
-        } ];
+          linkConfig = optionalAttrs ( i.macAddress != null) {
```
Comment (rnhmjoj) [inline nixos/modules/tasks/network-interfaces-systemd.nix:97]:
```suggestion
          linkConfig = optionalAttrs (i.macAddress != null) {
```

Just a nitpick

File: nixos/modules/tasks/network-interfaces.nix [hunk 3]
```diff
@@ -1031,6 +1031,11 @@ in
         message = ''
           Temporary addresses are only needed when IPv6 is enabled.
         '';
+      })) ++ (forEach interfaces (i: {
+        assertion = (!i.virtual) || i.virtualType != "tun" || i.macAddress == null;
```
[the remark this answers is not in the document]
Comment (flokli) [reply nixos/modules/tasks/network-interfaces.nix:1035]: But that wouldn't be the same - setting a mac adress for "tap" interfaces is perfectly fine.

I changed it to
```
assertion = (i.virtual && i.virtualType == "tun") -> i.macAddress == null;
```

File: nixos/modules/tasks/network-interfaces-systemd.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/tasks/network-interfaces-systemd.nix:97]: perfectly right :-) Thanks, done.

File: nixos/modules/tasks/network-interfaces.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (rnhmjoj) [reply nixos/modules/tasks/network-interfaces.nix]: > setting a mac adress for "tap" interfaces is perfectly fine.

Ah, sorry I didn't know. I ever never used these virtual interfaces.

**Review by rnhmjoj — APPROVED** [review]

**Review by lheckemann — APPROVED** [review]

**Comment by rnhmjoj** [comment]
I run all tests again on x86_64 and aarch64. Everything is passing.

**Comment by rnhmjoj** [comment]
Backported the fix in 897182cdaf5.

**Pull request merged**